model-engine/model_engine_server/db/base.py

```diff
@@ -87,97 +95,168 @@ def get_engine_url(env: Optional[str] = None, read_only: bool = True, sync: bool
+    # but also shouldn't get rid of as many errors e.g. 95% compared to 99.9%)
+    # option is to try pool_recycle = something kinda short e.g. a minute
+    # pool_pre_ping=True seems to not increase latency by very much
+    # (I profiled 2.7 ms -> 3.3 ms on GET model_bundles/)
+    # but hopefully should completely eliminate
+    # any of the postgres connection errors we've been seeing.
+
+    pg_engine = create_engine(
+        get_engine_url(read_only=False, sync=True, reset_expiration_timestamp=True),
+        echo=False,
+        future=True,
+        pool_pre_ping=True,
+        pool_size=20,
+        max_overflow=30,
+    )
+    pg_engine_read_only = create_engine(
+        get_engine_url(read_only=True, sync=True),
+        echo=False,
+        future=True,
+        pool_pre_ping=True,
+        pool_size=20,
+        max_overflow=30,
+    )
+    pg_engine_async = create_async_engine(
+        get_engine_url(read_only=False, sync=False),
+        echo=False,
+        future=True,
+        pool_pre_ping=True,
+        pool_size=20,
+        max_overflow=30,
+    )
+    pg_engine_read_only_async = create_async_engine(
+        get_engine_url(read_only=True, sync=False),
+        echo=False,
+        future=True,
+        pool_pre_ping=True,
+        pool_size=20,
+        max_overflow=30,
+    )
+    pg_engine_async_null_pool = create_async_engine(
+        get_engine_url(read_only=False, sync=False),
+        echo=False,
+        future=True,
+        poolclass=NullPool,
+        pool_pre_ping=True,
+    )
+
+    global _Session
+    global _SessionReadOnly
+    global _SessionAsync
+    global _SessionAsyncNullPool
+    global _SessionReadOnlyAsync
+
+    _Session = sessionmaker(autocommit=False, autoflush=False, bind=pg_engine)
+    _SessionReadOnly = sessionmaker(autocommit=False, autoflush=False, bind=pg_engine_read_only)
+    _SessionAsync = async_scoped_session(
+        session_factory=async_sessionmaker(
+            autocommit=False,
+            autoflush=False,
+            bind=pg_engine_async,
+            expire_on_commit=False,
+        ),
+        scopefunc=asyncio.current_task,
+    )
+    _SessionAsyncNullPool = async_scoped_session(
+        session_factory=async_sessionmaker(
+            autocommit=False,
+            autoflush=False,
+            bind=pg_engine_async_null_pool,
+            expire_on_commit=False,
+        ),
+        scopefunc=asyncio.current_task,
+    )
+    _SessionReadOnlyAsync = async_scoped_session(
+        async_sessionmaker(
+            autocommit=False,
+            autoflush=False,
+            bind=pg_engine_read_only_async,
+            expire_on_commit=False,
+        ),
+        scopefunc=asyncio.current_task,
+    )
+
+
+refresh_sessions()
+
+
+def get_session():
+    global _Session
+    global database_credential_expiration_timestamp
+
+    if time.time() > database_credential_expiration_timestamp:
```

Comment:
qq do we want a `- epsilon` here? although I'm not sure how long it takes the credentials to change/expire